ord_schema/templating_test.py

```diff
@@ -106,6 +104,44 @@ def test_bad_placeholders(self):
         with self.assertRaisesRegex(ValueError, r'\$conversion\$ not found'):
             templating.generate_dataset(template_string, df)
 
+    def test_missing_values(self):
+        # Build a template reaction.
+        reaction = reaction_pb2.Reaction()
+        input1 = reaction.inputs['one']
+        input1_component1 = input1.components.add()
+        input1_component1.identifiers.add(value='CCO', type='SMILES')
+        input1_component1.mass.value = 1.2
+        input1_component1.mass.units = reaction_pb2.Mass.GRAM
+        input1_component2 = input1.components.add()
+        input1_component2.is_limiting = True
+        input1_component2.identifiers.add(value='c1ccccc1', type='SMILES')
+        input1_component2.volume.value = 3.4
+        input1_component2.volume.units = reaction_pb2.Volume.LITER
+        input2 = reaction.inputs['two']
+        input2_component1 = input2.components.add()
+        input2_component1.identifiers.add(value='COO', type='SMILES')
+        input2_component1.mass.value = 5.6
+        input2_component1.mass.units = reaction_pb2.Mass.GRAM
+        outcome = reaction.outcomes.add()
+        outcome.conversion.value = 75
+        template_string = text_format.MessageToString(reaction)
+        template_string = template_string.replace('value: "CCO"',
+                                                  'value: "$smiles$"')
+        template_string = template_string.replace('value: 5.6', 'value: $mass$')
+        # Build a spreadsheet and test for proper edits.
+        df = pd.DataFrame([{'smiles': 'CN'}, {'mass': 1.5}])
+        dataset = templating.generate_dataset(template_string, df)
+        expected_dataset = dataset_pb2.Dataset()
+        reaction1 = expected_dataset.reactions.add()
+        reaction1.CopyFrom(reaction)
+        reaction1.inputs['one'].components[0].identifiers[0].value = 'CN'
+        del reaction1.inputs['two']  # No components after amount removal.
+        reaction2 = expected_dataset.reactions.add()
+        reaction2.CopyFrom(reaction)
+        del reaction2.inputs['one'].components[0]  # No indentifiers.
+        reaction2.inputs['two'].components[0].mass.value = 1.5
+        self.assertEqual(dataset, expected_dataset)
```

Comment:
It took me a very long time to understand this test case, admittedly. Would you be able to change the construction of your `df` template to construct from dictionaries? This will make it clearer that there are two partially-defined reactions: `{'smiles': ['CN', ''], 'mass': ['', 1.5]}` where maybe the empty string mass should be a NaN? Even better would be to construct a multiline .csv and read from a tempfile so it exactly mimics how the file might be parsed by Pandas.

Reply:
I'll create a CSV to mimic reality; good suggestion.